cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,139 @@
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.sourcecontrol.CommitMeta;
+import io.cdap.cdap.sourcecontrol.RepositoryManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * In-Memory implementation for {@link SourceControlOperationRunner}.
+ * Runs all git operation inside calling service.
+ */
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
+  private final RepositoryManager repositoryManager;
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(@Assisted RepositoryManager repositoryManager) {
+    this.repositoryManager = repositoryManager;
+  }
+
+
+  @Override
+  public PushAppsResponse push(List<ApplicationDetail> appsToPush, CommitMeta commitDetails)
+    throws PushFailureException {
+    try {
+      repositoryManager.cloneRemote();
+    } catch (Exception e) {
+      throw new PushFailureException("Failed to clone remote repository", e);
+    }
+    //TODO CDAP-20371: handle if appsToPush is too large
+    LOG.info("Pushing application configs for : {}", appsToPush.stream().map(ApplicationDetail::getName));
+
+    //TODO CDAP-20371: Add retry logic here in case the head at remote moved while we are doing push
+    modifyApplicationConfigAndPush(appsToPush, commitDetails);
+
+    List<PushAppResponse> responses = new ArrayList<>();
+
+    try {
+      for (ApplicationDetail detail : appsToPush) {
+        Path filePath = generateAndValidateApplicationConfigPath(detail.getName());
+        String gitFileHash = repositoryManager.getFileHash(filePath);
+        responses.add(new PushAppResponse(detail.getName(), detail.getAppVersion(), gitFileHash));
+      }
+    } catch (Exception e) {
+      LOG.debug("Failed to fetch push details: %s", e.getCause());
+      throw new PushFailureException("Failed to fetch push details", e);
+    }
+
+    return new PushAppsResponse(responses);
+  }
+
+  private void modifyApplicationConfigAndPush(List<ApplicationDetail> appsToPush, CommitMeta commitDetails)
+    throws PushFailureException {
+    List<Path> modifiedPaths = new ArrayList<Path>();
+    try {
+      // Create base dir if not exist.
+      // This already handles cases where one of the path prefix is not a directory
+      Files.createDirectories(repositoryManager.getBasePath());
+
+      for (ApplicationDetail appDetail : appsToPush) {
+        String appName = appDetail.getName();
+        Path configFilePath = generateAndValidateApplicationConfigPath(appName);
+        byte[] configData = GSON.toJson(appDetail).getBytes(StandardCharsets.UTF_8);
```

Comment:
should not change to byte[] and then write it again as that's 2 passes. There is a toJson() method that takes a Writer, should use that instead.

Reply:
updated with writer